File: com.github.tux2323.demo.chat.server/src/com/github/tux2323/demo/chatserver/internal/ChatServerImpl.scala
package com.github.tux2323.demo.chatserver.internal

import com.github.tux2323.demo.chatserver._
import org.osgi.util.tracker.ServiceTracker
import org.osgi.framework.ServiceReference

class ChatServerImpl extends ChatServer {

	override def login(username : String , password : String) : Session = {
		if(username == "Tester" && password == "test")
			new Session();
		else throw new RuntimeException("Invalid Login");
	}
	
	override def sendMessage(session : Session, msg : String) {
		val context = ApplicationContext.context;
		val serviceTracker = new ServiceTracker(context, classOf[ChatClient].getName(), null);
		serviceTracker.open();
		val services = serviceTracker.getServices();
		services.foreach(service => service.asInstanceOf[ChatClient].receiveMessage(msg));
		serviceTracker.close();
	}
	
	
}

File: com.github.tux2323.demo.chat.server/src/com/github/tux2323/demo/chat/server/internal/ChatServerImpl.java
package com.github.tux2323.demo.chat.server.internal;

import com.github.tux2323.demo.chat.server.ChatServer;
import com.github.tux2323.demo.chat.server.Session;

public class ChatServerImpl implements ChatServer {

	@Override
	public Session login(String username, String password) {
		Session session = new Session();
		session.setUsername(username);
		// TODO IDM Binding
		return session;
	}

}
